keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+|**Pod nomination and sorting**|Preferring pods with both `descheduler.alpha.kubernetes.io/request-evict-only` and `descheduler.alpha.kubernetes.io/eviction-in-progress` annotations.|Preferring pods that have a corresponding evacuation CR present.|
+|**Cache reconstruction**|When a descheduler gets restarted and a new internal caches (with tracked annotations) cleared the descheduler lists all pods and populate the cache with any pod that has both annotation (`descheduler.alpha.kubernetes.io/request-evict-only` and `descheduler.alpha.kubernetes.io/eviction-in-progress`) present. In case only the first annotation is present but the eviction request was already created, the update event handler will catch the second annotation addition and the cache gets synced. In the worst case the limit of max number of pods evicted gets exceeded.|The descheduler waits until all evacution CRs are synced.|
+
+Given the evacuation API is still a work in progress without any
+existing implementation, the annotation based eviction can be seen
+as an v1alpha1 implemenation of this proposal.
+Integration with the evacuation API as an v1alphaN or beta implementation.
+
+### Workflow Description
+
+This section outlines the workflow for enabling the proposed evictions
+in background that prevent workloads from being evicted right away
+without following additional eviction policies.
+
+#### Actors
+
+- Cluster Administrator: Responsible for configuring descheduling
+  policies and overseeing cluster operations.
+- User: Individuals or services running workloads within the cluster.
+- Descheduler: Component responsible for requesting pod evictions
+
+#### Workflow Steps
+
+1. Policy configuration
+   1. The cluster administrator configures the descheduler to enable the new functionality.
+2. A user deploys a workload through various nodes
+   1. A scheduler distributes the pods based on configured scheduling policies
+3. Eviction nomination
+   1. At the beginning of each descheduling cycle the descheduler increases
+      all the internal counters to take into account all pods that are subjects
+      to background eviction. Either by checking `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation,
+      list of evacuation requests or confronting the internal caches.
+   2. Each descheduling plugin nominates a set of pods to be evicted.
+   3. All the nominated pods are sorted based on the eviction-in-progress first priority.
+4. Pod eviction: the descheduler starts evicting nominated pods until a limit is reached
+   1. If an annotation based eviction is enabled:
+      1. If a pod is not annotated with `descheduler.alpha.kubernetes.io/request-evict-only` evict
+         the pod normally without additional handling.
```

Comment:
Can you please write the workflow twice? One for the `v1alpha1` and second for `v1alphaN or beta`?

Reply:
For example in the Evacuation case, Evacuation object can be created for any pod. Regardless if it supports evacuation or not. The evacuation controller and optional evacuator will ensure that the pod gets deleted eventually.

Reply:
Does it make sense for the descheduler to distinguish/prioritize pods that support evacuation?

Reply:
> Does it make sense for the descheduler to distinguish/prioritize pods that support evacuation?

Yes. It's preferable to evacuate pods which have higher chance of getting evicted at the end. Would presence of `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}` signal such a support?

Reply:
Sounds good. This can be implemented as an improvement. Though, every pod should have the same chance of getting evacuated/evicted.